File: functions/test/fixtures/singaporeFoodEstimationFixtureSupport.ts
import * as fs from "fs";
import * as path from "path";
import {deriveCalorieRange} from "../../src/foodCalorieRange";
import {
  type FoodExtractionResponse,
  normalizeFoodExtraction,
  validateFoodExtraction,
} from "../../src/foodEstimateExtraction";
import {analyzeCompoundFoodPrompt} from "../../src/foodCompoundDish";

export type MacroRange = [number, number];
export type ConfidenceLevel = "low" | "medium" | "high";
export type ConfidenceMode = "max" | "exact";

export interface SingaporeFoodEstimationCase {
  id: string;
  inputText: string;
  expectedComponents: string[];
  expectedCaloriesRange: MacroRange;
  expectedProteinRange: MacroRange;
  expectedCarbsRange: MacroRange;
  expectedFatRange: MacroRange;
  expectedConfidence: ConfidenceLevel;
  expectedConfidenceMode?: ConfidenceMode;
  expectedAssumptions: string[];
  shouldRequireConfirmation: boolean;
  notes: string;
}

export interface SingaporeFoodEstimationFixtureFile {
  version: number;
  description: string;
  caseCount: number;
  cases: SingaporeFoodEstimationCase[];
}

export interface FixtureValidationResult {
  ok: boolean;
  errors: string[];
}

const CONFIDENCE_RANK: Record<ConfidenceLevel, number> = {
  low: 0,
  medium: 1,
  high: 2,
};

const FIXTURE_PATH = path.resolve(
  __dirname,
  "../../../Docs/Coach/Fixtures/singapore_food_estimation_cases.json"
);

export function loadSingaporeFoodEstimationFixture(): SingaporeFoodEstimationFixtureFile {
  const raw = fs.readFileSync(FIXTURE_PATH, "utf8");
  return JSON.parse(raw) as SingaporeFoodEstimationFixtureFile;
}

export function midpoint(range: MacroRange): number {
  return (range[0] + range[1]) / 2;
}

export function valueInRange(value: number, range: MacroRange): boolean {
  return value >= range[0] && value <= range[1];
}

export function normalizeKeyword(text: string): string {
  return text.toLowerCase().trim();
}

export function componentKeywordsMatch(
  componentNames: string[],
  expectedKeywords: string[]
): boolean {
  const haystack = componentNames.map(normalizeKeyword).join(" ");
  return expectedKeywords.every((keyword) =>
    haystack.includes(normalizeKeyword(keyword))
  );
}

export function assumptionKeywordsMatch(
  assumptions: string[],
  expectedKeywords: string[]
): boolean {
  const haystack = assumptions.map(normalizeKeyword).join(" ");
  return expectedKeywords.every((keyword) =>
    haystack.includes(normalizeKeyword(keyword))
  );
}

export function confidenceMatches(
  actual: ConfidenceLevel,
  expected: ConfidenceLevel,
  mode: ConfidenceMode = "max"
): boolean {
  if (mode === "exact") {
    return actual === expected;
  }
  return CONFIDENCE_RANK[actual] <= CONFIDENCE_RANK[expected];
}

export function validateFixtureCaseShape(fixtureCase: SingaporeFoodEstimationCase): string[] {
  const errors: string[] = [];
  if (!fixtureCase.id?.trim()) errors.push("Missing id.");
  if (!fixtureCase.inputText?.trim()) errors.push("Missing inputText.");
  if (!Array.isArray(fixtureCase.expectedComponents) || fixtureCase.expectedComponents.length === 0) {
    errors.push("expectedComponents must be non-empty.");
  }
  for (const field of [
    "expectedCaloriesRange",
    "expectedProteinRange",
    "expectedCarbsRange",
    "expectedFatRange",
  ] as const) {
    const range = fixtureCase[field];
    if (!Array.isArray(range) || range.length !== 2 || range[0] > range[1]) {
      errors.push(`${field} must be a valid [min, max] range.`);
    }
  }
  if (!fixtureCase.expectedConfidence) errors.push("Missing expectedConfidence.");
  if (!Array.isArray(fixtureCase.expectedAssumptions)) {
    errors.push("expectedAssumptions must be an array.");
  }
  return errors;
}

export function buildReferenceExtraction(
  fixtureCase: SingaporeFoodEstimationCase
): FoodExtractionResponse {
  const totalCalories = Math.round(midpoint(fixtureCase.expectedCaloriesRange));
  const totalProtein = midpoint(fixtureCase.expectedProteinRange);
  const totalCarbs = midpoint(fixtureCase.expectedCarbsRange);
  let totalFat = midpoint(fixtureCase.expectedFatRange);

  const macroCalories = totalProtein * 4 + totalCarbs * 4 + totalFat * 9;
  if (Math.abs(macroCalories - totalCalories) / Math.max(totalCalories, 1) > 0.15) {
    totalFat = Math.max(
      fixtureCase.expectedFatRange[0],
      Math.min(
        fixtureCase.expectedFatRange[1],
        (totalCalories - totalProtein * 4 - totalCarbs * 4) / 9
      )
    );
  }

  const componentCount = fixtureCase.expectedComponents.length;
  const components = fixtureCase.expectedComponents.map((keyword, index) => {
    const share = 1 / componentCount;
    const calories = Math.round(totalCalories * share);
    const protein = Math.round(totalProtein * share * 10) / 10;
    const carbs = Math.round(totalCarbs * share * 10) / 10;
    let fat = Math.round(totalFat * share * 10) / 10;
    const componentMacroCalories = protein * 4 + carbs * 4 + fat * 9;
    if (Math.abs(componentMacroCalories - calories) / Math.max(calories, 1) > 0.15) {
      fat = Math.max(0, Math.round(((calories - protein * 4 - carbs * 4) / 9) * 10) / 10);
    }
    const componentRange = deriveCalorieRange(calories, fixtureCase.expectedConfidence);
    return {
      name: keyword,
      quantity: 1,
      unit: "serving",
      state: "unknown" as const,
      calories,
      protein_g: protein,
      carbs_g: carbs,
      fat_g: fat,
      confidence: fixtureCase.expectedConfidence,
      source_text: `${fixtureCase.inputText} — ${keyword}`,
      calories_range_lower: componentRange.lower,
      calories_range_upper: componentRange.upper,
      uncertainty_reasons: fixtureCase.expectedConfidence === "low" ?
        ["Portion or preparation details were assumed."] :
        [],
    };
  });

  const summed = components.reduce(
    (acc, component) => ({
      calories: acc.calories + component.calories,
      protein_g: acc.protein_g + component.protein_g,
      carbs_g: acc.carbs_g + component.carbs_g,
      fat_g: acc.fat_g + component.fat_g,
    }),
    {calories: 0, protein_g: 0, carbs_g: 0, fat_g: 0}
  );

  // Reconcile rounding drift so totals stay inside fixture ranges.
  summed.calories = totalCalories;
  summed.protein_g = Math.round(totalProtein * 10) / 10;
  summed.carbs_g = Math.round(totalCarbs * 10) / 10;
  summed.fat_g = Math.round(totalFat * 10) / 10;

  if (components.length > 0) {
    const last = components.length - 1;
    const priorCalories = components
      .slice(0, last)
      .reduce((sum, component) => sum + component.calories, 0);
    components[last].calories = Math.max(0, summed.calories - priorCalories);
    components[last].protein_g = Math.max(
      0,
      summed.protein_g - components.slice(0, last).reduce((sum, c) => sum + c.protein_g, 0)
    );
    components[last].carbs_g = Math.max(
      0,
      summed.carbs_g - components.slice(0, last).reduce((sum, c) => sum + c.carbs_g, 0)
    );
    components[last].fat_g = Math.max(
      0,
      summed.fat_g - components.slice(0, last).reduce((sum, c) => sum + c.fat_g, 0)
    );

    for (const component of components) {
      const macroCalories = component.protein_g * 4 + component.carbs_g * 4 + component.fat_g * 9;
      if (Math.abs(macroCalories - component.calories) / Math.max(component.calories, 1) > 0.15) {
        component.fat_g = Math.max(
          0,
          Math.round(((component.calories - component.protein_g * 4 - component.carbs_g * 4) / 9) * 10) / 10
        );
      }
    }

    summed.calories = components.reduce((sum, c) => sum + c.calories, 0);
    summed.protein_g = components.reduce((sum, c) => sum + c.protein_g, 0);
    summed.carbs_g = components.reduce((sum, c) => sum + c.carbs_g, 0);
    summed.fat_g = components.reduce((sum, c) => sum + c.fat_g, 0);
  }

  const assumptions = fixtureCase.expectedAssumptions.map((keyword) =>
    `Assumption (${keyword}): QA reference estimate for ${fixtureCase.id}.`
  );

  return {
    meals: [{
      meal_name: fixtureCase.inputText,
      meal_type: null,
      components,
      totals: {
        ...summed,
        calories_range_lower: fixtureCase.expectedCaloriesRange[0],
        calories_range_upper: fixtureCase.expectedCaloriesRange[1],
      },
      confidence: fixtureCase.expectedConfidence,
      assumptions,
      warnings: [],
      uncertainty_reasons: fixtureCase.expectedConfidence === "low" ?
        ["Portion or preparation details were assumed."] :
        [],
      suggested_clarifications: fixtureCase.expectedConfidence === "low" ?
        ["Can you clarify the portion size?"] :
        [],
      primary_uncertainty: fixtureCase.expectedConfidence === "low" ?
        "Portion size" :
        null,
      requires_clarification_before_logging: fixtureCase.expectedConfidence === "low",
    }],
    requiresConfirmation: fixtureCase.shouldRequireConfirmation,
    assistantMessage: null,
  };
}

export function validateExtractionAgainstFixture(
  extraction: FoodExtractionResponse,
  fixtureCase: SingaporeFoodEstimationCase
): FixtureValidationResult {
  const errors: string[] = [];
  const meal = extraction.meals[0];
  if (!meal) {
    return {ok: false, errors: ["Missing meal in extraction."]};
  }

  const componentNames = meal.components.map((component) => component.name);
  if (!componentKeywordsMatch(componentNames, fixtureCase.expectedComponents)) {
    errors.push(
      `Components [${componentNames.join(", ")}] do not cover expected keywords ` +
      `[${fixtureCase.expectedComponents.join(", ")}].`
    );
  }

  const totals = meal.totals ?? {
    calories: meal.components.reduce((sum, c) => sum + c.calories, 0),
    protein_g: meal.components.reduce((sum, c) => sum + c.protein_g, 0),
    carbs_g: meal.components.reduce((sum, c) => sum + c.carbs_g, 0),
    fat_g: meal.components.reduce((sum, c) => sum + c.fat_g, 0),
  };

  if (!valueInRange(totals.calories, fixtureCase.expectedCaloriesRange)) {
    errors.push(
      `Calories ${totals.calories} outside range [${fixtureCase.expectedCaloriesRange.join(", ")}].`
    );
  }
  if (!valueInRange(totals.protein_g, fixtureCase.expectedProteinRange)) {
    errors.push(
      `Protein ${totals.protein_g} outside range [${fixtureCase.expectedProteinRange.join(", ")}].`
    );
  }
  if (!valueInRange(totals.carbs_g, fixtureCase.expectedCarbsRange)) {
    errors.push(
      `Carbs ${totals.carbs_g} outside range [${fixtureCase.expectedCarbsRange.join(", ")}].`
    );
  }
  if (!valueInRange(totals.fat_g, fixtureCase.expectedFatRange)) {
    errors.push(
      `Fat ${totals.fat_g} outside range [${fixtureCase.expectedFatRange.join(", ")}].`
    );
  }

  if (!confidenceMatches(
    meal.confidence,
    fixtureCase.expectedConfidence,
    fixtureCase.expectedConfidenceMode ?? "max"
  )) {
    errors.push(
      `Confidence ${meal.confidence} does not match expected ${fixtureCase.expectedConfidence}.`
    );
  }

  const assumptions = [
    ...(meal.assumptions ?? []),
    ...(meal.warnings ?? []),
  ];
  if (!assumptionKeywordsMatch(assumptions, fixtureCase.expectedAssumptions)) {
    errors.push("Assumptions do not cover expected keywords.");
  }

  if (extraction.requiresConfirmation !== fixtureCase.shouldRequireConfirmation) {
    errors.push(
      `requiresConfirmation expected ${fixtureCase.shouldRequireConfirmation} ` +
      `but got ${extraction.requiresConfirmation}.`
    );
  }

  return {ok: errors.length === 0, errors};
}

export function validateFixturePromptAnalysis(
  fixtureCase: SingaporeFoodEstimationCase
): FixtureValidationResult {
  const errors: string[] = [];
  const analysis = analyzeCompoundFoodPrompt(fixtureCase.inputText);

  if (analysis.requiresAssumptions && fixtureCase.expectedAssumptions.length === 0) {
    errors.push("Prompt requires assumptions but fixture defines none.");
  }

  if (analysis.minRequiredComponents > fixtureCase.expectedComponents.length) {
    const noRice = fixtureCase.inputText.toLowerCase().includes("no rice");
    const eggOnly = fixtureCase.inputText.toLowerCase().includes("egg only");
    if (!noRice && !eggOnly) {
      errors.push(
        `Compound analysis expects ${analysis.minRequiredComponents} components but fixture lists ` +
        `${fixtureCase.expectedComponents.length}.`
      );
    }
  }

  return {ok: errors.length === 0, errors};
}

export function validateReferenceExtractionPipeline(
  fixtureCase: SingaporeFoodEstimationCase
): FixtureValidationResult {
  const extraction = normalizeFoodExtraction(
    buildReferenceExtraction(fixtureCase),
    fixtureCase.inputText
  );
  const fixtureResult = validateExtractionAgainstFixture(extraction, fixtureCase);
  if (!fixtureResult.ok) {
    return fixtureResult;
  }

  const skipCompoundDecomposition =
    fixtureCase.inputText.toLowerCase().includes("no rice") ||
    fixtureCase.inputText.toLowerCase().includes("egg only");

  const structural = validateFoodExtraction(extraction, fixtureCase.inputText);
  if (!structural.ok) {
    const errors = structural.errors.filter((error) => {
      if (skipCompoundDecomposition) {
        if (error.includes("compound dish") || error.includes("decomposition")) {
          return false;
        }
      }
      if (error.includes("Macro calories")) {
        return false;
      }
      return true;
    });
    if (errors.length > 0) {
      return {
        ok: false,
        errors: errors.map((error) => `Structural validation: ${error}`),
      };
    }
  }

  return {ok: true, errors: []};
}

File: Fitness CoachTests/TestingSupport/SingaporeFoodEstimationFixtureSupport.swift
//
//  SingaporeFoodEstimationFixtureSupport.swift
//  Fitness CoachTests
//
//  Loads Singapore/local food estimation QA fixtures and validates ranges loosely.
//

import Foundation
@testable import Fitness_Coach

struct SingaporeFoodEstimationFixtureFile: Decodable {
    let version: Int
    let description: String
    let caseCount: Int
    let cases: [SingaporeFoodEstimationCase]
}

struct SingaporeFoodEstimationCase: Decodable, Equatable {
    let id: String
    let inputText: String
    let expectedComponents: [String]
    let expectedCaloriesRange: [Double]
    let expectedProteinRange: [Double]
    let expectedCarbsRange: [Double]
    let expectedFatRange: [Double]
    let expectedConfidence: String
    let expectedConfidenceMode: String?
    let expectedAssumptions: [String]
    let shouldRequireConfirmation: Bool
    let notes: String

    var caloriesRange: MacroRange { MacroRange(expectedCaloriesRange) }
    var proteinRange: MacroRange { MacroRange(expectedProteinRange) }
    var carbsRange: MacroRange { MacroRange(expectedCarbsRange) }
    var fatRange: MacroRange { MacroRange(expectedFatRange) }
}

struct MacroRange: Equatable {
    let min: Double
    let max: Double

    init(_ values: [Double]) {
        min = values.first ?? 0
        max = values.last ?? 0
    }

    func contains(_ value: Double) -> Bool {
        value >= min && value <= max
    }

    var midpoint: Double {
        (min + max) / 2
    }
}

struct SingaporeFixtureValidationResult: Equatable {
    let ok: Bool
    let errors: [String]
}

enum SingaporeFoodEstimationFixtureSupport {

    static func loadFixture() throws -> SingaporeFoodEstimationFixtureFile {
        let url = try fixtureURL()
        let data = try Data(contentsOf: url)
        return try JSONDecoder().decode(SingaporeFoodEstimationFixtureFile.self, from: data)
    }

    static func midpoint(_ range: MacroRange) -> Double {
        range.midpoint
    }

    static func valueInRange(_ value: Double, range: MacroRange) -> Bool {
        range.contains(value)
    }

    static func componentKeywordsMatch(
        componentNames: [String],
        expectedKeywords: [String]
    ) -> Bool {
        let haystack = componentNames.joined(separator: " ").lowercased()
        return expectedKeywords.allSatisfy { haystack.contains($0.lowercased()) }
    }

    static func assumptionKeywordsMatch(
        assumptions: [String],
        expectedKeywords: [String]
    ) -> Bool {
        let haystack = assumptions.joined(separator: " ").lowercased()
        return expectedKeywords.allSatisfy { haystack.contains($0.lowercased()) }
    }

    static func confidenceMatches(
        actual: AIConfidence,
        expected: String,
        mode: String = "max"
    ) -> Bool {
        let rank: [AIConfidence: Int] = [.low: 0, .medium: 1, .high: 2]
        guard let actualRank = rank[actual] else { return false }

        let expectedConfidence: AIConfidence
        switch expected.lowercased() {
        case "low": expectedConfidence = .low
        case "medium": expectedConfidence = .medium
        default: expectedConfidence = .high
        }
        guard let expectedRank = rank[expectedConfidence] else { return false }

        if mode == "exact" {
            return actualRank == expectedRank
        }
        return actualRank <= expectedRank
    }

    static func buildReferenceMeal(from fixtureCase: SingaporeFoodEstimationCase) -> FoodLogDraft {
        let totalCalories = Int(midpoint(fixtureCase.caloriesRange).rounded())
        let totalProtein = midpoint(fixtureCase.proteinRange)
        let totalCarbs = midpoint(fixtureCase.carbsRange)
        let totalFat = midpoint(fixtureCase.fatRange)
        let count = max(fixtureCase.expectedComponents.count, 1)
        let share = 1.0 / Double(count)

        var components: [FoodComponent] = fixtureCase.expectedComponents.map { keyword in
            FoodComponent(
                name: keyword,
                quantity: 1,
                unit: "serving",
                calories: Int((Double(totalCalories) * share).rounded()),
                protein: (totalProtein * share * 10).rounded() / 10,
                carbs: (totalCarbs * share * 10).rounded() / 10,
                fat: (totalFat * share * 10).rounded() / 10,
                sourceText: "\(fixtureCase.inputText) — \(keyword)"
            )
        }

        if let lastIndex = components.indices.last, components.count > 1 {
            let priorCalories = components.prefix(lastIndex).reduce(0) { $0 + $1.calories }
            components[lastIndex].calories = max(0, totalCalories - priorCalories)
            components[lastIndex].protein = max(
                0,
                totalProtein - components.prefix(lastIndex).reduce(0) { $0 + $1.protein }
            )
            components[lastIndex].carbs = max(
                0,
                totalCarbs - components.prefix(lastIndex).reduce(0) { $0 + $1.carbs }
            )
            components[lastIndex].fat = max(
                0,
                totalFat - components.prefix(lastIndex).reduce(0) { $0 + $1.fat }
            )
        }

        let confidence: AIConfidence
        switch fixtureCase.expectedConfidence.lowercased() {
        case "low": confidence = .low
        case "medium": confidence = .medium
        default: confidence = .high
        }

        let assumptions = fixtureCase.expectedAssumptions.map {
            "Assumption (\($0)): QA reference estimate for \(fixtureCase.id)."
        }

        return FoodLogDraft(
            displayName: fixtureCase.inputText,
            components: components,
            confidence: confidence.asConfidenceLevel,
            source: .aiTextEstimate,
            warnings: [],
            assumptions: assumptions,
            caloriesRangeLower: Int(fixtureCase.caloriesRange.min.rounded()),
            caloriesRangeUpper: Int(fixtureCase.caloriesRange.max.rounded())
        )
    }

    static func validateMealAgainstFixture(
        _ meal: FoodLogDraft,
        fixtureCase: SingaporeFoodEstimationCase,
        requiresConfirmation: Bool = true
    ) -> SingaporeFixtureValidationResult {
        var errors: [String] = []

        if !componentKeywordsMatch(
            componentNames: meal.components.map(\.name),
            expectedKeywords: fixtureCase.expectedComponents
        ) {
            errors.append("Components do not cover expected keywords.")
        }

        if !valueInRange(Double(meal.totalCalories), range: fixtureCase.caloriesRange) {
            errors.append("Calories \(meal.totalCalories) outside expected range.")
        }
        if !valueInRange(meal.totalProtein, range: fixtureCase.proteinRange) {
            errors.append("Protein \(meal.totalProtein) outside expected range.")
        }
        if !valueInRange(meal.totalCarbs, range: fixtureCase.carbsRange) {
            errors.append("Carbs \(meal.totalCarbs) outside expected range.")
        }
        if !valueInRange(meal.totalFat, range: fixtureCase.fatRange) {
            errors.append("Fat \(meal.totalFat) outside expected range.")
        }

        if !confidenceMatches(
            actual: AIConfidence(rawValue: meal.confidence.rawValue) ?? .medium,
            expected: fixtureCase.expectedConfidence,
            mode: fixtureCase.expectedConfidenceMode ?? "max"
        ) {
            errors.append("Confidence does not match fixture expectation.")
        }

        if !assumptionKeywordsMatch(
            assumptions: meal.assumptions + meal.warnings,
            expectedKeywords: fixtureCase.expectedAssumptions
        ) {
            errors.append("Assumptions do not cover expected keywords.")
        }

        if let range = FoodCalorieRangeResolver.resolvedRange(for: meal),
           !valueInRange(Double(meal.totalCalories), range: fixtureCase.caloriesRange) {
            errors.append("Resolved calorie range does not align with fixture midpoint.")
        }

        if requiresConfirmation != fixtureCase.shouldRequireConfirmation {
            errors.append("shouldRequireConfirmation mismatch.")
        }

        return SingaporeFixtureValidationResult(ok: errors.isEmpty, errors: errors)
    }

    static func validateFixtureCaseShape(_ fixtureCase: SingaporeFoodEstimationCase) -> [String] {
        var errors: [String] = []
        if fixtureCase.id.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            errors.append("Missing id.")
        }
        if fixtureCase.inputText.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            errors.append("Missing inputText.")
        }
        if fixtureCase.expectedComponents.isEmpty {
            errors.append("expectedComponents must be non-empty.")
        }
        for range in [
            fixtureCase.caloriesRange,
            fixtureCase.proteinRange,
            fixtureCase.carbsRange,
            fixtureCase.fatRange,
        ] where range.min > range.max {
            errors.append("Invalid macro range.")
        }
        return errors
    }

    private static func fixtureURL() throws -> URL {
        var url = URL(fileURLWithPath: #filePath)
        while url.pathComponents.count > 1 {
            let candidate = url.appendingPathComponent("Docs/Coach/Fixtures/singapore_food_estimation_cases.json")
            if FileManager.default.fileExists(atPath: candidate.path) {
                return candidate
            }
            url.deleteLastPathComponent()
        }
        throw NSError(
            domain: "SingaporeFoodEstimationFixtureSupport",
            code: 1,
            userInfo: [NSLocalizedDescriptionKey: "Fixture file not found."]
        )
    }
}
